pkg/controller/statusmanager/kube.go

```diff
@@ -0,0 +1,124 @@
+package statusmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/openshift/cluster-network-operator/pkg/apply"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Some silly types so that we can actually do testing
+
+type DaemonSetLister interface {
+	List(selector labels.Selector) (ret []*appsv1.DaemonSet, err error)
+}
+
+type DeploymentLister interface {
+	List(selector labels.Selector) (ret []*appsv1.Deployment, err error)
+}
+
+type StatefulSetLister interface {
+	List(selector labels.Selector) (ret []*appsv1.StatefulSet, err error)
+}
+
+type patchAnnotations struct {
+	Metadata md `json:"metadata"`
+}
+type md struct {
+	Annotations map[string]interface{} `json:"annotations"`
+}
+
+func makeAnnotationPatch(key string, value *string) ([]byte, error) {
+	patch := &patchAnnotations{
+		Metadata: md{
+			Annotations: map[string]interface{}{
+				key: value,
+			},
+		},
+	}
+	patchData, err := json.Marshal(&patch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create patch: %v", err)
+	}
+	return patchData, nil
+}
+
+func alreadySet(obj metav1.Object, key string, value *string) bool {
+	anno := obj.GetAnnotations()
+	existing, set := anno[key]
+	if value != nil && set && existing == *value {
+		return true
+	}
+	if !set && value == nil {
+		return true
+	}
+	return false
+}
+
+// setDSAnnotation sets an annotation on a daemonset; or unsets it if value is nil
+func (status *StatusManager) setDSAnnotation(obj *appsv1.DaemonSet, key string, value *string) error {
+	if alreadySet(obj, key, value) {
+		return nil
+	}
+	patchData, err := makeAnnotationPatch(key, value)
+	if err != nil {
+		return err
+	}
+
+	_, err = status.client.ClientFor(apply.GetClusterName(obj)).Kubernetes().AppsV1().DaemonSets(obj.Namespace).Patch(
+		context.TODO(),
+		obj.Name,
+		types.MergePatchType,
+		patchData,
+		metav1.PatchOptions{},
+	)
+	return err
+}
+
+// setSSAnnotation sets an annotation on a statefulset; or unsets it if value is nil
+func (status *StatusManager) setSSAnnotation(obj *appsv1.StatefulSet, key string, value *string) error {
```

Comment:
I know you're just copying the existing names but `SS` in particular is not obvious to me. None of these function names would really be too long if you just fully unabbreviated them...

Although also, given the whole dynamic client thing couldn't you just squash them into a single method? Or were we trying to get away from the dynamic client?

Reply:
There are several annoying effects in play here.

The first is that it's not possible to create an arbitrary label-scoped informer using the controller-runtime client (at least yet). One needs to manually create an informer. Assuming you don't want to pass Unstructured around, then you need to use the typed client to create the informers. Okay, fine, we create the informers ourselves and just wire it up. No big deal

The second problem is exclusively around testing. Basically, we offer 3 clients: the controller-runtime client, the Unstructured client, and the Structured client. Annoyingly, the fake clients **all have separate stores**. This means that the status controller's extensive test suite needs to use the same client for the same objects.

But now that I type this up, I think I have a way to make this problem go away, so it's good that you mentioned this.